generator/internal/service_generator.cc

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/service_generator.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/str_cat.h"
+#include "generator/internal/codegen_utils.h"
+#include "generator/internal/stub_generator.h"
+#include <google/api/client.pb.h>
+#include <google/protobuf/descriptor.h>
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+
+ServiceGenerator::ServiceGenerator(
```

Comment:
It doesn't look like this class needs to have a dep on `ServiceDescriptor`, nor `GeneratorContext`, nor the map of vars. I think this class only needs to have a dep on the `StubGenerator`s. Consider factoring those extra deps into a factory function for the StubGenerator, which will simplify this class and make it more flexible and easier to test. Specifically consider something like:

```cc
// stub_generator.h

// The implementation would look just like this current c'tor, but it would be in the *stub_generator* file.
std::unique_ptr<StubGenerator> MakeStubGenerator(
    ServiceDescriptor const& service, 
    GeneratorContext& context, 
    std::map<string, string> vars);
```

```cc
// service_generator.h
class ServiceGenerator {
 public:
  explicit ServiceGenerator(std::vector<std::unique_ptr<StubGenerator>> v);
  ...
};
```